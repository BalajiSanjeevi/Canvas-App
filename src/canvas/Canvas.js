import write from '../io/OutputWriter';
import {validateNumbers, validateCommandSetLength} from '../validators/Validator';

const EXPECTED_INPUT_LENGTH = 2;

const BOUNDARY_BORDER_PIXEL_LENGTH = 2;

const LEFT_AND_RIGHT_CANVAS_BORDER_CHAR = "|";

const TOP_AND_BOTTOM_CANVAS_BORDER_CHAR = "-";


/*
 * Canvas class is meant to hold the canvas context which will have the canvas boundary within which we can play around
 * and it has couple of function to help in drawing line and getting the context.
 */
export default class Canvas {

    /*
     * Validates the commandset for the required length and checks whether the command set is all numbers.
     */
    static validate = function (commandSet){
        return  validateCommandSetLength(commandSet, EXPECTED_INPUT_LENGTH) && validateNumbers(commandSet);
    };

    /*
     * Constructor to create the canvas object and initializes the canvas boundary.
     */
    constructor([width, height]){
        this._pixels = [[]];
        this._width = parseInt(width);
        this._height =  parseInt(height);
        this.initializeCanvasBoundary(this._height + BOUNDARY_BORDER_PIXEL_LENGTH, this._width + BOUNDARY_BORDER_PIXEL_LENGTH);
    }

    /*
     * Helps in initializing the canvas boundary.
     */
    initializeCanvasBoundary(height, width){
        this._pixels = [... new Array(height)].map((row, i) => {
            return [... new Array(width)].map((column, j) => {
                if(i === 0 || i === height - 1){
                    return TOP_AND_BOTTOM_CANVAS_BORDER_CHAR;
                } else if(j === 0 || j === width - 1) {
                    return LEFT_AND_RIGHT_CANVAS_BORDER_CHAR;
                }
                return " ";
            });
        });
    }

    /*
     * Exposes the pixels which contains the actual canvas boundary with the state of it.
     */
    get pixels(){
        return this._pixels;
    }

    /*
     *  Calls the write function to print all the rows and columns.
     */
    draw(){
        this._pixels.forEach((row) => {
            write(row.join(""));
        })
    }

    /*
     * Fills the character 'X' from the given startPoint and endPoint index in the given axisY row
     */
    drawHorizontalLine(startPoint, endPoint, axisY){
        // As per Fill function, it fills the character until < endPoint. As we need to include the last position like (<=endPoint). so adding 1 to the endpoint.
        this._pixels[axisY].fill('X', startPoint, endPoint + 1);
    }
    /*
     * Fills the character 'X' on each row from the given startPoint and endPoint index at the given axisX column index.
     */
    drawVerticalLine(startPoint, endPoint, axisX){
        for(var i = startPoint; i <= endPoint; i++){
            this._pixels[i][axisX] = 'X';
        }
    }
}