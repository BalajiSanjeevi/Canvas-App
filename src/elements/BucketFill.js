import Pixel from '../canvas/Pixel';
import bucketValidate from '../validators/BucketFillValidator';

export default class BucketFill {

    constructor([axisX, axisY, fillChar]){
        this._startPoint = new Pixel(axisX, axisY);
        this._fillChar = fillChar;
    }

    static validate = bucketValidate;

    /*
     * Flood fill, also called seed fill, is an algorithm that determines the area connected to a given node in a multi-dimensional array.
     * It is used in the "bucket" fill tool of paint programs to fill connected. The below function has been implemented based on
     * Recursive flood fill with 4 directions. https://en.wikipedia.org/wiki/Flood_fill
     */
    floodFill(canvas, pixel){
        if(canvas.pixels[pixel.axisY][pixel.axisX] === " "){
            canvas.pixels[pixel.axisY][pixel.axisX] = this._fillChar;
            this.floodFill(canvas, new Pixel(pixel.axisX + 1, pixel.axisY));
            this.floodFill(canvas, new Pixel(pixel.axisX - 1, pixel.axisY));
            this.floodFill(canvas, new Pixel(pixel.axisX, pixel.axisY + 1));
            this.floodFill(canvas, new Pixel(pixel.axisX, pixel.axisY - 1));
        }
    }
    /*
     * Function which is being called to fill and draw the pixels for bucket fill command.
     */
    draw(canvas){
        this.floodFill(canvas, this._startPoint);
        canvas.draw();
    }
}