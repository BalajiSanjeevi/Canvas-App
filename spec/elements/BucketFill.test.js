import BucketFill from '../../src/elements/BucketFill';
import Canvas from '../../src/canvas/Canvas';

jest.mock('../../src/io/OutputWriter');

describe("BucketFill", () => {
    let bucketFill, canvas;

    beforeEach(() => {
        canvas = new Canvas([10, 4]);
    });

    it("should match snapshot", () => {
        canvas.drawHorizontalLine(1, 4, 2);
        canvas.drawVerticalLine(2, 4, 4);
        bucketFill = new BucketFill([5, 2, '*']);
        bucketFill.draw(canvas);
        expect(canvas.pixels).toMatchSnapshot();
    });

});