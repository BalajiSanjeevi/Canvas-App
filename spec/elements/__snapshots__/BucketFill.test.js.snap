// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BucketFill should match snapshot 1`] = `
------------
|**********|
|XXXX******|
|   X******|
|   X******|
------------
`;
